File: grace-payment-integration/src/router.ts
import { PaymentClient, MerchantAuthenticationClient, types } from 'hs-paylib';
import { CYBERSOURCE_CONFIG, PAYPAL_CONFIG, ADYEN_CONFIG } from './config';

// Pre-initialized clients for connection pool reuse
const cybersourceClient = new PaymentClient(CYBERSOURCE_CONFIG);
const paypalClient = new PaymentClient(PAYPAL_CONFIG);
const paypalAuthClient = new MerchantAuthenticationClient(PAYPAL_CONFIG);
const adyenClient = new PaymentClient(ADYEN_CONFIG);

export type ConnectorName = 'cybersource' | 'paypal' | 'adyen';

export interface RouteResult {
  connector: ConnectorName;
  client: PaymentClient;
}

/**
 * Routes payments based on currency and amount:
 * - EUR -> PayPal
 * - USD amount > 100 (minor units > 10000) -> Cybersource
 * - USD amount <= 100 (minor units <= 10000) -> Adyen
 */
export function routePayment(currency: types.Currency, minorAmount: number): RouteResult {
  if (currency === types.Currency.EUR) {
    return { connector: 'paypal', client: paypalClient };
  }

  // USD routing by amount threshold (100 USD = 10000 minor units)
  if (minorAmount > 10000) {
    return { connector: 'cybersource', client: cybersourceClient };
  }

  return { connector: 'adyen', client: adyenClient };
}

/**
 * Obtain a PayPal access token for card payments.
 */
export async function getPayPalAccessToken(): Promise<{
  token: string;
  expiresInSeconds: number;
}> {
  const tokenResponse = await paypalAuthClient.createServerAuthenticationToken({
    merchantAccessTokenId: `token_${Date.now()}`,
    connector: types.Connector.PAYPAL,
    testMode: true,
  });

  if (!tokenResponse.accessToken?.value) {
    throw new Error('Failed to obtain PayPal access token');
  }

  return {
    token: tokenResponse.accessToken.value,
    expiresInSeconds: Number(tokenResponse.expiresInSeconds || 0),
  };
}

export function parseCurrency(currency: string): types.Currency {
  const value = (types.Currency as unknown as Record<string, number>)[currency];
  if (value === undefined) {
    throw new Error(`Unsupported currency: ${currency}`);
  }
  return value;
}

File: payment-server/router.py
"""
Currency-based payment router.

Routes payments to the correct connector based on currency:
  USD -> PayPal
  EUR -> Cybersource
"""

from payments.generated import sdk_config_pb2


CURRENCY_CONNECTOR_MAP = {
    "USD": "paypal",
    "EUR": "cybersource",
}


def get_connector_for_currency(currency: str) -> str:
    """Return connector name for the given currency, or raise ValueError."""
    currency = currency.upper()
    connector = CURRENCY_CONNECTOR_MAP.get(currency)
    if not connector:
        supported = ", ".join(CURRENCY_CONNECTOR_MAP.keys())
        raise ValueError(
            f"Unsupported currency: {currency}. Supported: {supported}"
        )
    return connector


def get_config_for_currency(
    currency: str,
    configs: dict[str, sdk_config_pb2.ConnectorConfig],
) -> tuple[str, sdk_config_pb2.ConnectorConfig]:
    """Return (connector_name, config) for the given currency."""
    connector = get_connector_for_currency(currency)
    return connector, configs[connector]
